### PR TITLE
Feature/37857 Pin availability changes

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**
- [x]  Admin backend pin generation changes
- [x]  Pupil api taking into account new `pinValidFrom` field

### CODE REVIEW

File: admin/services/data-access/check-form-allocation.data.service.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ checkFormAllocationDataService.sqlFindByIdsHydrated = function (ids) {
       chk.isLiveCheck as check_isLiveCheck,
       pin.val as pupil_pin,
       cp.pinExpiresAt as pupil_pinExpiresAt,
+      cp.pinValidFrom as pupil_pinValidFrom,
```
Comment (GeorgeChatzigiannis) [inline admin/services/data-access/check-form-allocation.data.service.js]: I am not quite sure if this is really needed yet.

File: admin/data/sql/migrations/20200210153747.do.mod-check-table-type-and-spcreatecheck.sql [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  pupil_id int,
+  checkForm_id int,
+  checkWindow_id int,
+  isLiveCheck bit,
+  pinExpiresAt datetimeoffset(3),
+  pinValidFrom datetimeoffset(3),
+  school_id int
+);
+
+GO
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153747.do.mod-check-table-type-and-spcreatecheck.sql]: Should be a new migration if you need to call `GO`

File: admin/data/sql/migrations/20200210153747.undo.mod-check-table-type-and-spcreatecheck.sql [hunk 3]
```diff
@@ -0,0 +1,139 @@
+-- DROP PROCEDURE IN ORDER TO RECREATE THE CHECK TABLE TYPE
+DROP PROCEDURE IF EXISTS [mtc_admin].[spCreateChecks];
+DROP TYPE IF EXISTS [mtc_admin].[checkTableType];
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153747.undo.mod-check-table-type-and-spcreatecheck.sql]: Don't you need to recreate the TableType before defining the Procedure?

File: admin/services/prepare-check.service.js [hunk 4]
```diff
@@ -88,7 +88,8 @@ function constructPreparedCheck (check) {
     config: check.config,
     createdAt: moment(),
     pinExpiresAt: moment(check.pupil.pinExpiresAt),
-    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt'], check.pupil),
+    pinValidFrom: moment(check.pupil.pinValidFrom),
+    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt', 'pinValidFrom'], check.pupil),
```
Comment (jon-shipley) [inline admin/services/prepare-check.service.js]: Doesn't this duplicate the `pinValidFrom` field in the pupil object too?

File: admin/data/sql/migrations/20200210153747.undo.mod-check-table-type-and-spcreatecheck.sql (hunk 3 above)
Comment (GeorgeChatzigiannis) [reply admin/data/sql/migrations/20200210153747.undo.mod-check-table-type-and-spcreatecheck.sql]: @jon-shipley migrations have been updated please re-review.

File: admin/services/prepare-check.service.js [hunk 5]
```diff
@@ -87,8 +87,9 @@ function constructPreparedCheck (check) {
     checkCode: check.pupil.checkCode,
     config: check.config,
     createdAt: moment(),
-    pinExpiresAt: moment(check.pupil.pinExpiresAt),
-    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt'], check.pupil),
+    pinExpiresAtUtc: moment(check.pupil.pinExpiresAtUtc),
+    pinValidFromUtc: moment(check.pupil.pinValidFromUtc),
+    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAtUtc', 'pinValidFromUtc'], check.pupil),
```
Comment (GeorgeChatzigiannis) [inline admin/services/prepare-check.service.js]: @jon-shipley I believe we need to omit the new column `pinValidFromUtc` as we don't want this to exist in the pupil object of the prepared check.

File: admin/data/sql/migrations/20200210153633.do.update-sp-create-checks-pt-1.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE mtc_admin.checkPin ADD pinValidFromUtc [datetimeoffset](3) NOT NULL;
+EXEC sp_rename 'mtc_admin.checkPin.pinExpiresAt', pinExpiresAtUtc, 'COLUMN';
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153633.do.update-sp-create-checks-pt-1.sql]: Surely this is best left alone?

File: admin/data/sql/migrations/20200210153633.undo.update-sp-create-checks-pt-1.sql [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            -- Create the check
+            INSERT INTO [mtc_admin].[check]
+            (pupil_id, checkForm_id, checkWindow_id, isLiveCheck)
+            VALUES (@pupilId, @checkFormId, @checkWindowId, @isLiveCheck);
+
+            -- Get the check.id we just inserted
+            SET @checkId = SCOPE_IDENTITY();
+
+            -- Assign a pin to the check
+            INSERT INTO [mtc_admin].[checkPin] (school_id, check_id, pinExpiresAt, pin_id)
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153633.undo.update-sp-create-checks-pt-1.sql]: This appears to be a very good reason to leave it alone - it should be pinExpiresAtUtc according to the migration above.

Just leave it - it's a minor change that has lots of knock-on effects.

File: admin/data/sql/migrations/20200210153633.undo.update-sp-create-checks-pt-1.sql [hunk 8]
```diff
@@ -0,0 +1,135 @@
+CREATE OR ALTER PROCEDURE [mtc_admin].[spCreateChecks]
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153633.undo.update-sp-create-checks-pt-1.sql]: I don't see what changed here.   Can you explain?

File: admin/data/sql/migrations/20200210153747.undo.update-sp-create-checks-pt-2.sql [hunk 9]
```diff
@@ -0,0 +1,2 @@
+EXEC sp_rename 'mtc_admin.checkPin.pinExpiresAtUtc', pinExpiresAt, 'COLUMN';
+ALTER TABLE mtc_admin.checkPin DROP COLUMN pinValidFromUtc;
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200210153747.undo.update-sp-create-checks-pt-2.sql]: The undo here does not re-create the spCreateChecks proc which it must.

What it ought to do here is:

1) Drop and recreate the TVP
2) Drop and recreate the SP spCreateChecks

These ought to be separate migrations.

File: admin/data/sql/migrations/20200211161754.undo.update-sp-create-checks-pt-3.sql [hunk 10]
```diff
@@ -0,0 +1,12 @@
+-- DROP PROCEDURE IN ORDER TO RECREATE THE CHECK TABLE TYPE
+DROP PROCEDURE IF EXISTS [mtc_admin].[spCreateChecks];
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200211161754.undo.update-sp-create-checks-pt-3.sql]: This undo migration does not undo the migration.

File: pupil-api/src/services/redis-pupil-auth.service.spec.ts [hunk 11]
```diff
@@ -79,6 +85,46 @@ describe('redis-pupil-auth.service', () => {
...
+    redisServiceMock.get = jest.fn(async (key: string) => {
+      return expectedPayload
+    })
+    const schoolPin = 'abc12def'
+    const pupilPin = '5678'
+    const payload = await sut.authenticate(schoolPin, pupilPin)
+    expect(payload).toBeUndefined()
+  })
+
+  test('the check payload should not be returned if check pinExpiresAtUtc timestamp is before current date', async () => {
```
Comment (jon-shipley) [inline pupil-api/src/services/redis-pupil-auth.service.spec.ts]: Could you rephrase this test? Perhaps 'authorisation denied if the pin has expired' ?

File: pupil-api/src/services/redis-pupil-auth.service.spec.ts [hunk 12]
```diff
@@ -65,10 +67,14 @@ describe('redis-pupil-auth.service', () => {
   })
 
   test('the check payload should be returned if item found in cache', async () => {
```
Comment (jon-shipley) [inline pupil-api/src/services/redis-pupil-auth.service.spec.ts]: This test, _as described_, is not a valid test, as the pin must pass time-based validity checks.

File: pupil-api/src/services/redis-pupil-auth.service.spec.ts [hunk 13]
```diff
@@ -79,6 +85,46 @@ describe('redis-pupil-auth.service', () => {
     expect(payload).toEqual(expectedPayload)
   })
 
+  test('the check payload should not be returned if check pinValidFromUtc timestamp is after current date', async () => {
```
Comment (jon-shipley) [inline pupil-api/src/services/redis-pupil-auth.service.spec.ts]: Perhaps you could rephrase this test to save us all a headache 🤕 😁 

Suggest 'authorisation is denied if the pin is not yet valid'?

**Comment by jon-shipley** [comment]
@GuyHarwood @GeorgeChatzigiannis - you have forgotten to change all views using `pinExpiresAt`!

File: admin/data/sql/migrations/20200210153633.do.update-sp-create-checks-pt-1.sql (hunk 6 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20200210153633.do.update-sp-create-checks-pt-1.sql]: @GeorgeChatzigiannis you have misunderstood the request - we want the object in the payload to have the Utc postfix, not the database column

File: admin/data/sql/migrations/20200211161754.do.update-sp-create-checks-pt-3.sql [hunk 14]
```diff
@@ -0,0 +1,137 @@
+CREATE OR ALTER PROCEDURE [mtc_admin].[spCreateChecks]
+@TVP [mtc_admin].[CheckTableType] READONLY
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/20200211161754.do.update-sp-create-checks-pt-3.sql]: as discussed, please remove as no requirement to store valid from in database

File: admin/services/check-start.service/check-start.service.js [hunk 15]
```diff
@@ -17,14 +17,17 @@ const checkFormService = require('../check-form.service')
...
-const pinGenerationService = require('../pin-generation.service')
+const pinTimestampService = require('../pin-timestamp.service')
 const prepareCheckService = require('../prepare-check.service')
 const queueNameService = require('../queue-name-service')
 const sasTokenService = require('../sas-token.service')
 const redisCacheService = require('../data-access/redis-cache.service')
 const redisKeyService = require('../redis-key.service')
 const oneMonthInSeconds = 2592000
 
+const fourPmToday = moment().startOf('day').add(16, 'hours')
```
Comment (jon-shipley) [inline admin/services/check-start.service/check-start.service.js]: This is an expression that will resolve to a value, marked as const.  How does it get updated for the next day?

File: admin/services/pin-generation.service.js [hunk 16]
```diff
@@ -17,31 +17,12 @@ const bannedWords = [
   'dim'
 ]
 
-const fourPmToday = () => moment().startOf('day').add(16, 'hours')
```
Comment (jon-shipley) [inline admin/services/pin-generation.service.js]: In the old implementation, `fourPmToday` was a function.

File: admin/services/pin-generation.service.js [hunk 17]
```diff
@@ -17,31 +17,12 @@ const bannedWords = [
...
-  } else {
-    toReturn = fourPmToday()
-  }
-  if (schoolTimezone) {
-    // needed to parse the date in the specified timezone and convert to utc for storing
-    toReturn = moment.tz(dateService.formatIso8601WithoutTimezone(toReturn), schoolTimezone).utc()
-  }
-  return toReturn
-}
+const fourPmToday = moment().startOf('day').add(16, 'hours')
```
Comment (jon-shipley) [inline admin/services/pin-generation.service.js]: Duplication

File: admin/services/pin-generation.service.js [hunk 18]
```diff
@@ -80,7 +61,7 @@ pinGenerationService.generateSchoolPassword = school => {
     wordsArray[pinGenerationService.generateCryptoRandomNumber(0, wordsArray.length - 1)]
   const numberCombination = randomGenerator.getRandom(2, chars)
   const newPin = `${firstRandomWord}${numberCombination}${secondRandomWord}`
-  const newExpiry = pinExpiryTime(school.timezone)
+  const newExpiry = pinTimestampService.generatePinTimestamp(config.OverridePinExpiry, endOfDay, fourPmToday, school.timezone)
```
Comment (jon-shipley) [inline admin/services/pin-generation.service.js]: There seem to be too many arguments here.  Surely, the first 3 aren't actually needed?

It can get the config itself.
endfOfDay and fourPmToday are derived from time, although passing them in as args may assist in testing.  Is that why?

File: admin/services/prepare-check.service.js [hunk 19]
```diff
@@ -3,27 +3,33 @@
...
 
+const config = require('../config')
 const logger = require('./log.service').getLogger()
+const pinTimestampService = require('./pin-timestamp.service')
 const prepareCheckDataService = require('./data-access/prepare-check.data.service')
 const redisCacheService = require('./data-access/redis-cache.service')
 const redisKeyService = require('./redis-key.service')
 const redisService = require('./data-access/redis-cache.service')
 
+const startOfDay = moment().startOf('day')
```
Comment (jon-shipley) [inline admin/services/prepare-check.service.js]: Again, can't see how this will work after day 1.

File: admin/services/pin-generation.service.js (hunk 16 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pin-generation.service.js]: @jon-shipley resolved

File: admin/services/pin-generation.service.js (hunk 18 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pin-generation.service.js]: as discussed, i am resolving this.

File: admin/services/pin-generation.service.js (hunk 17 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pin-generation.service.js]: these consts are required in 2 different services.
1)pin-generation.service for generating school password's `pinExpiresAt`
2)check-start.service for generating check's `pinExpiresAt`

Comment (GuyHarwood) [reply admin/services/pin-generation.service.js]: and they will also be making an appearance in the school pin generator function :)

File: admin/services/check-start.service/check-start.service.js [hunk 20]
```diff
@@ -17,14 +17,17 @@ const checkFormService = require('../check-form.service')
...
-const pinGenerationService = require('../pin-generation.service')
+const pinTimestampService = require('../pin-timestamp.service')
 const prepareCheckService = require('../prepare-check.service')
 const queueNameService = require('../queue-name-service')
 const sasTokenService = require('../sas-token.service')
 const redisCacheService = require('../data-access/redis-cache.service')
 const redisKeyService = require('../redis-key.service')
 const oneMonthInSeconds = 2592000
 
+const fourPmToday = () => moment().startOf('day').add(16, 'hours')
```
Comment (GuyHarwood) [inline admin/services/check-start.service/check-start.service.js]: unless these are used anywhere else, i suggest moving them into the pinTimestampService

File: admin/services/pin-timestamp.service.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
+const moment = require('moment')
+
+const dateService = require('../services/date.service')
+
+const pinTimestampService = {}
```
Comment (GuyHarwood) [inline admin/services/pin-timestamp.service.js]: pinTimestamp seems like a vague name

File: admin/services/prepare-check.service.js [hunk 22]
```diff
@@ -80,14 +86,17 @@ function secondsBetweenNowAndPinExpiryTime (pinExpiry) {
 /**
  * Return an entity suitable for inserting into the `preparedCheck` table
  * @param {object} check
+ * @param {string} schoolTimezone
  * @return {object}
  */
-function constructPreparedCheck (check) {
+function constructPreparedCheck (check, schoolTimezone) {
+  const pinValidFromUtc = pinTimestampService.generatePinTimestamp(config.OverridePinValidFrom, startOfDay(), eightAmToday(), schoolTimezone)
```
Comment (GuyHarwood) [inline admin/services/prepare-check.service.js]: i would suggest `pinValidFromGenerator` or something more specific to its actual purpose

File: admin/services/check-start.service/check-start.service.js (hunk 20 above)
Comment (GeorgeChatzigiannis) [reply admin/services/check-start.service/check-start.service.js]: these are needed in the caller as the implementation service can expect at the moment either `fourPmToday`(different calls in different services for pupil pin and school password) or `eightAmToday` (pupil pin only).

File: admin/services/pin-generation.service.js [hunk 23]
```diff
@@ -80,7 +61,7 @@ pinGenerationService.generateSchoolPassword = school => {
     wordsArray[pinGenerationService.generateCryptoRandomNumber(0, wordsArray.length - 1)]
   const numberCombination = randomGenerator.getRandom(2, chars)
   const newPin = `${firstRandomWord}${numberCombination}${secondRandomWord}`
-  const newExpiry = pinExpiryTime(school.timezone)
+  const newExpiry = pinValidityGeneratorService.generatePinTimestamp(config.OverridePinExpiry, endOfDay(), fourPmToday(), school.timezone)
```
Comment (jon-shipley) [inline admin/services/pin-generation.service.js]: I have to admit that I am confused by the meaning of `pinValidityGeneratorService`.  Can validity be generated?  Also, in comp-sci, `generator` often has a specific meaning, which is not meant here.

If we remove all the extra words here, I see `pinService`.  I think this is simpler and more direct.

Comment (GeorgeChatzigiannis) [reply admin/services/pin-generation.service.js]: moved to `pinService`
